File: Ashutosh-Singh/2D-Array/Reshape-the-Matrix.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
    vector<vector<int>> matrixReshape(vector<vector<int>>& mat, int r, int c) {
        int n = mat.size();      
        int m = mat[0].size();    

        // If reshaping is not possible, return the original matrix
        if (n * m != r * c) {
            return mat; 
        }

        vector<vector<int>> ans(r, vector<int>(c));
        int row = 0, col = 0;

        // Reshaping the matrix
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                ans[row][col] = mat[i][j];
                col++;
                if (col == c) { 
                    col = 0;
                    row++;
                }
            }
        }

        return ans;
    }
};

int main() {
    Solution solution;
    
    int n, m, r, c;
    
    // Input matrix dimensions
    cout << "Enter number of rows (n): ";
    cin >> n;
    cout << "Enter number of columns (m): ";
    cin >> m;
    
    vector<vector<int>> mat(n, vector<int>(m));
    
    // Input matrix elements
    cout << "Enter the elements of the matrix:" << endl;
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < m; j++) {
            cin >> mat[i][j];
        }
    }
    
    // Input target dimensions for reshaping
    cout << "Enter number of rows (r) for reshaped matrix: ";
    cin >> r;
    cout << "Enter number of columns (c) for reshaped matrix: ";
    cin >> c;

    // Get reshaped matrix
    vector<vector<int>> result = solution.matrixReshape(mat, r, c);

    // Output the reshaped matrix
    cout << "Reshaped matrix: " << endl;
    for (const auto& row : result) {
        for (int val : row) {
            cout << val << " ";
        }
        cout << endl;
    }

    return 0;
}

File: [Manya]/Leetcode/2D-Array/Reshape_the_matrix.java
import java.util.Scanner;

class Solution {
    public int[][] matrixReshape(int[][] mat, int r, int c) {
        int r1 = mat.length;
        int c1 = mat[0].length;

        if (r1 * c1 != r * c) {
            return mat;
        }

        int[][] reshaped = new int[r][c];

        int r2 = 0;
        int c2 = 0;

        for (int i = 0; i < r1; i++) {
            for (int j = 0; j < c1; j++) {
                reshaped[r2][c2] = mat[i][j];
                c2++; 
                if (c2 == c) { 
                    c2 = 0; 
                    r2++;  
                }
            }
        }
        return reshaped;
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        
        System.out.print("Enter the number of rows in the original matrix: ");
        int r1 = scanner.nextInt();
        System.out.print("Enter the number of columns in the original matrix: ");
        int c1 = scanner.nextInt();
        
        int[][] mat = new int[r1][c1];
        System.out.println("Enter the elements of the matrix:");
        for (int i = 0; i < r1; i++) {
            for (int j = 0; j < c1; j++) {
                mat[i][j] = scanner.nextInt();
            }
        }
        
        System.out.print("Enter the number of rows for the reshaped matrix: ");
        int r = scanner.nextInt();
        System.out.print("Enter the number of columns for the reshaped matrix: ");
        int c = scanner.nextInt();
        
        Solution solution = new Solution();
        int[][] result = solution.matrixReshape(mat, r, c);
        
        System.out.println("Reshaped Matrix:");
        for (int i = 0; i < result.length; i++) {
            for (int j = 0; j < result[i].length; j++) {
                System.out.print(result[i][j] + " ");
            }
            System.out.println();
        }
        
        scanner.close();
    }
}
